### PR TITLE
fix Issue 15865 - std.file.copy(from,to) deletes the file if from and to specify the same file

### PR DESCRIPTION
**Pull request by CyberShadow**
https://issues.dlang.org/show_bug.cgi?id=15865

Stat the target fd and check for sameness before truncating the file.

### CODE REVIEW

File: std/file.d [hunk 1]
```diff
@@ -3196,9 +3207,9 @@ unittest
     import std.algorithm, std.file; // issue 14817
     auto t1 = deleteme, t2 = deleteme~"2";
     scope(exit) foreach (t; [t1, t2]) if (t.exists) t.remove();
-    write(t1, "1");
+    write(t1, "11");
```
Comment (schveiguy) [inline std/file.d]: I'm not complaining, but just wondering why this change happened?

**Comment by schveiguy** [comment]
LGTM.

File: std/file.d (hunk 1 above)
Comment (CyberShadow) [inline std/file.d]: To test that the file is actually truncated, i.e. that we don't just overwrite the first `fileSize(source)` bytes.

Comment (schveiguy) [inline std/file.d]: Ah, OK. The importance comes later.

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on
